### DISCUSSION TITLE
c++ 20 module library cannot be found

### DISCUSSION THREAD

**Discussion by hexne** [xmake-io/xmake#5825] [Q&A] [answered]
I have a library of cppm files that I want to use in xmake, so I fork the xrepo repository and add an xmake.lua file, but that still doesn't work in xmake. Here's some information：
Content of repo xmake.lua
https://paste.ubuntu.com/p/tdtrkk55kM/

new project xmake.lua :
![Image_1731598741696.png](https://github.com/user-attachments/assets/201cd184-0d98-4818-b905-0b64828ce93a)

build && error message: 
![Image_1731598743060.png](https://github.com/user-attachments/assets/c4545f44-27b0-407e-b48d-4fc6e5b9c92d)

local repo info：
![Image_1731598745234.png](https://github.com/user-attachments/assets/1e7c4c92-a41e-4e88-a36a-b3693ed1631d)

This seems to be similar to example https://github.com/xmake-io/xmake/blob/3a7d68de17f19afadc2f7eec698d54861ecd729d/tests/projects/c++/modules/packages, but I don't understand why this library can't be used normally.

@Arthapz 
thanks.

**Comment by Arthapz** [comment, marked as answer]
ur package is not correct and u missed the configs = { module = true } in ur add_requires

here the correct version

```lua
-- nelib package
package("nelib")
    set_kind("library")
    set_homepage("https://github.com/hexne/nelib")
    set_description("A commonly used tool collection library")

    add_urls("https://github.com/hexne/NeLib.git")
    add_versions("2024.11.10", "6ec115b4cc87556809567413d1ab94c303179737")

    add_configs("modules", {description = "Build with C++20 modules support.", default = false, type = "boolean"})
    add_configs("header_only", {description = "Build as a headeronly library.", default = false, type = "boolean"})
    set_policy("package.cmake_generator.ninja", true)

    add_deps("opencv")

    on_load(function (package)
        if not package:config("modules") then
            package:add("deps", "cmake")
            if package:config("header_only") then
                package:set("kind", "library", {headeronly = true})
            end
        end
    end)

    on_install("linux", function (package)
        if not package:config("modules") then
            os.cp("*.h", package:installdir("include"))
            if not package:config("header_only") then
                io.replace("CMakeLists.txt", "NOT GLM_DISABLE_AUTO_DETECTION", "FALSE")
                local configs = {"-DGLM_BUILD_TESTS=OFF"}
                table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
                import("package.tools.cmake").build(package, configs)
                os.cp("**.pcm", package:installdir("include"))
                os.cp("**.a", package:installdir("lib"))
            end
        else
            io.writefile("xmake.lua", [[ 
                add_requires("opencv")
                target("nelib")
                    add_packages("opencv")
                    set_kind("$(kind)")
                    set_languages("c++20")
                    add_headerfiles("./**.h", {outputdir = 'nelib'})
                    add_includedirs(".")
                    add_files("**.cppm", {public = true})
            ]])
            import("package.tools.xmake").install(package)
        end

    end)


    on_test(function (package)
        assert(package:check_cxxsnippets({test = [[
            void test() {
                ;
            }
        ]]}, {configs = {languages = "c++20", includes = "nelib/tools.h"}}))
    end)
```

```lua
-- xmake.lua
add_repositories("nelib test-repo")
add_requires("nelib", {configs = {modules = true}})

target("foo")
    add_files("src/*.cpp")
    add_packages("nelib")
    set_policy("build.c++.modules", true)
```

(but your library seems to not build :D)
```
error: @programdir/core/main.lua:329: @programdir/actions/build/main.lua:148: @programdir/modules/async/runjobs.lua:325: ...ramdir/rules/c++/modules/modules_support/gcc/builder.lua:64: @programdir/modules/core/tools/gcc.lua:938: ThreadPool.cppm: In destructor 'nl::ThreadPool::~ThreadPool()':
ThreadPool.cppm:59:9: error: 'for_each' was not declared in this scope
   59 |         for_each(work_.begin(), work_.end(), std::mem_fn(&std::thread::join)); //等待
      |         ^~~~~~~~
ThreadPool.cppm: At global scope:
ThreadPool.cppm:15:8: warning: not writing module 'ThreadPool' due to errors
   15 | export module ThreadPool;
      |        ^~~~~~
stack traceback:
```

**Reply by hexne** [reply]
ohh！！, thinks very much, I have modified my library so that they are warnings instead of errors, but it still won't compile under libc++, but I will fix it.
![f0fb948f-d201-4c32-aeb4-2424a8c83b9d.png](https://github.com/user-attachments/assets/a37a57a6-b28f-43df-b085-e5ad124b660a)

**Reply by Arthapz** [reply]
--toolchain=clang --runtimes="c++_shared" if you are not using macOS, otherwise clang will use libstdc++

**Reply by hexne** [reply]
Sorry to bother you again, but I solved one problem with --runtimes=c++_shared, but I couldn't solve the other one:  
When using`` xmake f -vD --toolchain=clang --runtimes=c++_shared -c``, I get this error message:
```
error: @programdir/core/main.lua:329: @programdir/actions/build/main.lua:148: @programdir/modules/async/runjobs.lua:325: ...mdir/rules/c++/modules/modules_support/clang/builder.lua:95: @programdir/modules/core/tools/gcc.lua:938: Time.cppm:112:10: error: no type named 'jthread' in namespace 'std'; did you mean 'thread'?
  112 |     std::jthread thread_;
      |     ~~~~~^~~~~~~
      |          thread
/usr/bin/../include/c++/v1/__thread/thread.h:148:33: note: 'thread' declared here
  148 | class _LIBCPP_EXPORTED_FROM_ABI thread {
      |                                 ^
Time.cppm:117:28: error: no member named 'jthread' in namespace 'std'
  117 |     Timer() : thread_(std::jthread(&Timer::run,this)) {  }
      |                       ~~~~~^
Time.cppm:121:24: error: no member named 'get_stop_token' in 'std::thread'
  121 |         return thread_.get_stop_token().stop_requested();
      |                ~~~~~~~ ^
3 errors generated.

```
This error looks like I didn't enable C++20 or forgot to include<thread>.
I can make sure that this Time.cppm file includes <thread>, 
![图片](https://github.com/user-attachments/assets/bddb75aa-a0cb-4770-9fac-d011a0733661)  
The new xmkae-repo xmake.lua file has the following contents:
https://github.com/hexne/xmake-repo/blob/dev/packages/n/nelib/xmake.lua

At the same time, you set the language standard to c++20 in xmake.lua, and the project that uses this module also sets the language standard to c++20 in xmake.lua, but it still indicates that there is no jthread in std.    
As the previous screenshot shows, when I used ``xmake f -vD --toolchain=clang -c``, it would compile and run correctly, but with a warning.  
![图片](https://github.com/user-attachments/assets/18f8a2e2-f5a5-43b4-9dbb-f64ee5cb3222)

In addition, I have already installed libc++, if you need additional information, please let me know, there is too much information generated by the compilation and I am not sure what is really useful for you.  
![图片](https://github.com/user-attachments/assets/337ae05a-f354-44b3-bcd6-c3a7db296ef9)

**Reply by hexne** [reply]
I'm sorry to bother you, but I think I've found the reason.
in [xxx ](https://libcxx.llvm.org/Status/Cxx20.html) have:
![图片](https://github.com/user-attachments/assets/2fda87a1-96a6-4b3e-9784-6c6b332436ec)
